notebook4e.py

```diff
@@ -1156,3 +1159,251 @@ def plot_pomdp_utility(utility):
+
+def path_states(node):
+    "The sequence of states to get to this node."
+    if node in (cutoff, failure, None): 
+        return []
+    return path_states(node.parent) + [node.state]
+
+def straight_line_distance(A, B):
+    "Straight-line distance between two points."
+    return sum(abs(a - b)**2 for (a, b) in zip(A, B)) ** 0.5
+
+
+class PriorityQueue:
+    """A queue in which the item with minimum f(item) is always popped first."""
+
+    def __init__(self, items=(), key=lambda x: x):
+        self.key = key
+        self.items = [] # a heap of (score, item) pairs
+        for item in items:
+            self.add(item)
+
+    def add(self, item):
+        """Add item to the queue."""
+        pair = (self.key(item), item)
+        heapq.heappush(self.items, pair)
+
+    def pop(self):
+        """Pop and return the item with min f(item) value."""
+        return heapq.heappop(self.items)[1]
+
+    def top(self): return self.items[0][1]
+
+    def __len__(self): return len(self.items)
+
+
+class GridProblem(Problem):
+    """Finding a path on a 2D grid with obstacles. Obstacles are (x, y) cells."""
+
+    def __init__(self, initial=(0, 0), goal=(9, 9), obstacles=(), **kwds):
+        Problem.__init__(self, initial=initial, goal=goal,
+                         obstacles=set(obstacles) - {initial, goal}, **kwds)
+
+    directions = [(-1, -1), (0, -1), (1, -1),
+                  (-1, 0),           (1,  0),
+                  (-1, +1), (0, +1), (1, +1)]
+    
+    def step_cost(self, s, action, s1): return straight_line_distance(s, s1)
+    
+    def h(self, node): return straight_line_distance(node.state, self.goal)
+
+    def result(self, state, action): 
+        "Both states and actions are represented by (x, y) pairs."
+        return action if action not in self.obstacles else state
+    
+    def draw_walls(self):
+        self.obstacles |= {(i, -2) for i in range(-2, self.width+4)}
+        self.obstacles |= {(i, self.height+4) for i in range(-2, self.width+4)}
+        self.obstacles |= {(-2, j) for j in range(-2, self.height+5)}
+        self.obstacles |= {(self.width+4, j) for j in range(-2, self.height+5)}
+
+    def actions(self, state):
+        """You can move one cell in any of `directions` to a non-obstacle cell."""
+        x, y = state
+        return {(x + dx, y + dy) for (dx, dy) in self.directions} - self.obstacles
+
+
+def random_lines(X=range(0, 10), Y=range(10), N=20, lengths=range(1, 3)):
+    """The set of cells in N random lines of the given lengths."""
+    result = set()
+    for _ in range(N):
+        x, y = random.choice(X), random.choice(Y)
+        dx, dy = random.choice(((0, 1), (1, 0)))
+        result |= line(x, y, dx, dy, random.choice(lengths))
+    return result
+
+
+def line(x, y, dx, dy, length):
+    """A line of `length` cells starting at (x, y) and going in (dx, dy) direction."""
+    return {(x + i * dx, y + i * dy) for i in range(length)}
+
+
+class AnimateProblem(GridProblem):
```

Comment:
Isn't there a way to get just the `AnimateProblem`  function in `notebook.py`? I think the rest do not fit here, except `random_lines` and `line`.

Reply:
We need methods from the class `GridProblem` (which in turn is a subclass of `Problem`) and removing the class will result in the addition of the same methods in `AnimateProblem` class. I know that it seems odd to repeat the same code, but we can get away with that if these functions and classes are just utilities for internal use. What do you think?

Reply:
Ok, I took a shot at it and removed almost all the duplicated code. Runs fine with a few warnings from `matplotlib` animation API.

Reply:
Oh! The warning was because the solution initially is empty and matplotlib is not able to find the required arguments to draw... I think it's fine.